src/tfvc/commands/undo.ts

```diff
@@ -0,0 +1,107 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+"use strict";
+
+import * as path from "path";
+import { TeamServerContext} from "../../contexts/servercontext";
+import { IArgumentProvider, IExecutionResult, ITfvcCommand } from "../interfaces";
+import { TfvcError } from "../tfvcerror";
+import { ArgumentBuilder } from "./argumentbuilder";
+import { CommandHelper } from "./commandhelper";
+
+/**
+ * This command undoes the changes to the files passed in.
+ * It returns a list of all files undone.
+ * undo [/recursive] <itemSpec>...
+ */
+//TODO: Add an Undo All?
+export class Undo implements ITfvcCommand<string[]> {
+    private _serverContext: TeamServerContext;
+    private _itemPaths: string[];
+
+    public constructor(serverContext: TeamServerContext, itemPaths: string[]) {
+        if (!itemPaths || itemPaths.length === 0) {
+            throw TfvcError.CreateArgumentMissingError("itemPaths");
+        }
+        this._serverContext = serverContext;
+        this._itemPaths = itemPaths;
+    }
+
+    public GetArguments(): IArgumentProvider {
+        return new ArgumentBuilder("undo", this._serverContext)
+            .AddAll(this._itemPaths);
+    }
+
+    public GetOptions(): any {
+        return {};
+    }
+
+    /**
+     * Example of output
+     * Undoing edit: file1.java
+     * Undoing add: file2.java
+     */
+    public async ParseOutput(executionResult: IExecutionResult): Promise<string[]> {
+        //TODO: (Undo All) What if we've been passed 5 files and the 3rd has no pending changes?
+        if (CommandHelper.HasError(executionResult, "No pending changes were found for ")) {
```

Comment:
Your TODO is valid. You may have errors and "Undoing edit" lines interlaced. You need to skip those errors and just continue.